README.md

# photobiologyInOut <img src="man/figures/logo.png" align="right" width="120" />

[![CRAN
version](https://www.r-pkg.org/badges/version-last-release/photobiologyInOut)](https://cran.r-project.org/package=photobiologyInOut)
[![cran
checks](https://cranchecks.info/badges/worst/photobiologyInOut)](https://cran.r-project.org/web/checks/check_results_photobiologyInOut.html)
[![R build
status](https://github.com/aphalo/photobiologyInOut/workflows/R-CMD-check/badge.svg)](https://github.com/aphalo/photobiologyInOut/actions)

Package ‘**photobiologyInOut**’ provides functions for importing
spectral data from diverse sources including instrument-specific files,
spectral data output by solar-radiation simulation models, and files
from on-line data repositories. It also includes functions for
conversion of spectral data from/to classes defined in other R packages.
Package ‘**photobiologyInOut**’ complements other packages in the
‘**r4photobiology suite**’ by allowing reading and writing “foreign”
spectral data as well as reading data saved from data loggers.

Data files from **spectrometers** from the following suppliers are
currently supported: Avantes, LI-COR, Macam Photonics, and Ocean Optics.

Data files from **data loggers** from the following suppliers are
currently supported: Campbell Scientific and YoctoPuce.

Data files from output by **radiation transfer models**: libRadtran and
TUV.

Data objects of classes defined in **R packages**: ‘hyperSpec’,
‘colorSpec’ and ‘pavo’.

Data files downloaded from **repositories of spectral data**: ASTER
(NASA’s ECOSTRESS Spectral Library) and FReD (Floral Reflectance
Database).

Developing a package like this is a never-ending task as I have only a
limited sample of output files for testing and formats are quite
variable. The functions may not work with different software or firmware
versions used for acquiring spectral data from instruments. Even the
format of files can depend on the current locale and operating system.

This package is part of a suite of R packages for photobiological
calculations described at the
[r4photobiology](https://www.r4photobiology.info) web site.

## Warning

**The functions in this package work with the example files I have
access to for testing, but they may not work with your own files as file
formats vary.**

**PLEASE, BE VERY CAREFUL WHEN USING THIS PACKAGE. DO CHECK THAT UNITS
USED IN THE IMPORTED FILE ARE THOSE EXPECTED BY THESE FUNCTIONS AND THAT
THE VALUES IN THE RETRIEVED DATA ARE THOSE EXPECTED!**

*If the functions in this package do not work with your files, they
hopefully will be useful as examples for developing your own functions.
If you develop new functions or improve the existing ones, please, do
contribute them back to this project.*

## Installation

Installation of the most recent stable version from CRAN:

``` r
install.packages("photobiologyInOut")
```

Installation of the current unstable version from GitHub:

``` r
# install.packages("devtools")
devtools::install_github("aphalo/photobiologyInOut")
```

## Documentation

HTML documentation is available at
(<https://docs.r4photobiology.info/photobiologyInOut/>), including a
*User Guide*.

News on updates to the different packages of the ‘r4photobiology’ suite
are regularly posted at (<https://www.r4photobiology.info/>).

Two articles introduce the basic ideas behind the design of the suite
and its use: Aphalo P. J. (2015)
(<https://doi.org/10.19232/uv4pb.2015.1.14>) and Aphalo P. J. (2016)
(<https://doi.org/10.19232/uv4pb.2016.1.15>).

A book is under preparation, and the draft is currently available at
(<https://leanpub.com/r4photobiology/>).

A handbook written before the suite was developed contains useful
information on the quantification and manipulation of ultraviolet and
visible radiation: Aphalo, P. J., Albert, A., Björn, L. O., McLeod, A.
R., Robson, T. M., & Rosenqvist, E. (Eds.) (2012) Beyond the Visible: A
handbook of best practice in plant UV photobiology (1st ed., p. xxx +
174). Helsinki: University of Helsinki, Department of Biosciences,
Division of Plant Biology. ISBN 978-952-10-8363-1 (PDF),
978-952-10-8362-4 (paperback). PDF file available from
(<https://doi.org/10.31885/9789521083631>).

## Contributing

Pull requests, bug reports, and feature requests are welcome at
(<https://github.com/aphalo/photobiologyInOut>). Contribution of example
data files that could be supported in future versions will be very much
appreciated.

## Citation

If you use this package to produce scientific or commercial
publications, please cite according to:

``` r
citation("photobiologyInOut")
#> 
#> To cite package 'photobiologyInOut' in publications, please use:
#> 
#>   Aphalo, Pedro J. (2015) The r4photobiology suite. UV4Plants Bulletin,
#>   2015:1, 21-29. DOI:10.19232/uv4pb.2015.1.14
#> 
#> A BibTeX entry for LaTeX users is
#> 
#>   @Article{,
#>     author = {Pedro J. Aphalo},
#>     title = {The r4photobiology suite},
#>     journal = {UV4Plants Bulletin},
#>     volume = {2015},
#>     number = {1},
#>     pages = {21-29},
#>     year = {2015},
#>     doi = {10.19232/uv4pb.2015.1.14},
#>   }
```

## License

© 2015-2021 Pedro J. Aphalo (<pedro.aphalo@helsinki.fi>). Released under
the GPL, version 2 or greater. This software carries no warranty of any
kind.


NEWS
TODO
====

Add support for StellarNet, Brewer, Wasatch and Bentham spectrometers (examples
of data files are welcome for these and other instruments).
Add a generic reader of files into a collection of spectra, or update current
functions to handle character vectors of paths.
Add test cases for new functions for exchange of foreign data (currently tested 
through examples in vignette) (URGENT otherwise bugs creep-in easily).

photobiologyInOut 0.4.23 (2021-04-09)
=====================================

* Add parameter ... to read_csi_data().
* Track various changes in the tidyverse that deprecated function used in this
package.

photobiologyInOut 0.4.22-1 (2020-04-09)
=======================================

* Fix two bad test cases that passed with 'photobiology' (== 0.10.0) but not
with 'photobiology' (>= 0.10.1).

photobiologyInOut 0.4.22 (2020-04-02)
=======================================

* Revise read_fmi2mspct() adding flexibility to cope with variation in date
and time formats used in headers of spectra.
* Update for compatibility with 'tibble' (>= 3.0.0).
* Update for compatibility with 'photobiology' (>= 0.10.0) which also provides 
compatibility with 'dplyr' (>= 1.0.0).
* Depends now on 'photobiology' (>= 0.10.0).

photobiologyInOut 0.4.21-1 (2020-01-10)
=======================================

* Package code unchanged.
* Update test cases for 'photobiology' (>= 0.9.30).
* Depends now on 'photobiology' (>= 0.9.30).

photobiologyInOut 0.4.21 (2019-08-18)
=======================================

* Track changes in 'tidyr' (= 1.0.0).
* Move packages 'pavo' and 'hyperSpec' to suggests.

photobiologyInOut 0.4.20 (2019-06-08)
=======================================

* Track changes in 'photobiology' (= 0.9.28) and 'dplyr' (= 0.8.1).
* Set "how.measured" attribute in file import functions.

photobiologyInOut 0.4.19 (2018-12-11)
=======================================

* Add checks for non-ASCII characters in headers.
* Track code breaking change in 'readr' (>= 1.3.0).

photobiologyInOut 0.4.18 (2018-11-22)
=======================================

* Track code breaking change in 'readr' (>= 1.2.0).
* Fix bug in read_oo_pidata() that was preventing reading of Ocean Optics 
"Raspberry Pi"  Flame files with a sequence header line at the top.
* Add decoding of file header to extract integration time and number of scans 
to read_oo_pidata().
* Improve decoding of file header to extract instrument settings in 
read_oo_jazdata().

photobiologyInOut 0.4.17 (2018-09-27)
=====================================

* Add new function to read transmittance and reflectance files from Ocean 
Optics' Jaz modular spectrometer.
* Fix bug in decoding of time and date in functions read_fmi2mspct(), 
read_tuv_usrout() and read_qtuv_txt(). Decoding failed under R 3.4.4 (oldrel) 
but worked under R 3.5.1 due to functions from 'lubridate' returning POSIXlt 
values instead of POSIXct values. Now we force conversion to POSIXct.

photobiologyInOut 0.4.16 (2018-08-30)
=====================================

Revise read_fmi_dat() to extract date from file header and save the file header
to the spectral object as a comment.
Add read_fmi2mspct() for reading spectral irradiance.
Bug fix: decoding of dates in fmi import functions for daily data not always 
returned POSIXct objects triggering errors in downstream code.

photobiologyInOut 0.4.15 (2018-08-10)
=====================================

Add read_yoctopuce_csv() to read CSV files from YoctoPuce modules.
Add read_qtuv_txt() to read spectral data output files from the on-line
Quick TUV calculator.
Add read_tuv_usrout2mspct() wrapper of read_tuv_usrout() and subset2mspct().
Revise read_tuv_usrout() with new formal parameter ozone.du and fix a bug in
the setting of "where.measured" attribute.
Revise all read functions so that they store the header of the imported file as
a vector of character strings to attribute "file.header".

photobiologyInOut 0.4.14 (2018-04-03)
=====================================

Implement class coercion methods ("as." methods) to complement earlier functions,
tracking 'photobiology' 0.9.20.
Move coercion methods from and to matrix objects to 'photobiology' 0.9.20.
Update to track API changes in 'colorSpec' 0.7-3.
Translate vignette to Rmarkdown.

photobiologyInOut 0.4.13
========================

Add function read_FReD_csv() (read files flower reflectance database).
Add function read_ASTER_txt() (read files ASTER materials reflectance database).
Revise function read_licor_prn() to accept transmittance and reflectance spectra
in addition to energy and photon irradiance spectra.
Improve handling of 'label' by using only the base name of files and appending
user supplied label text.
Change location of external data used in vignettes.

photobiologyInOut 0.4.12
========================

Fix CITATION file.
Add function read_uvspec_disort(), rename read_libradtran_vesa() to
read_uvspec_disort_vesa().

photobiologyInOut 0.4.11
========================

Edit read_licor_prn() to work around a bug in readr::read_table() that results
in missread data (wrong numbers).
Add read_csi_dat() to read data saved from modern Campbell Scientific loggers.

photobiologyInOut 0.4.10
========================

Make compatible with upcoming version of package 'lubridate' (pull request by
vspinu NA).
Add function read_avaspec_xls() which can import spectral data from Excel
files exported from Avantes software.
Fix code to avoid partial matches of function parameter names (tz -> tzone).

photobiologyInOut 0.4.9
=======================

Fix LaTeX error in vignette.

photobiologyInOut 0.4.8
=======================

Replace a couple of NA returned by read fucntions with empty spectral objects.
Improve handling of strict.range to avoid spurious warnings.

photobiologyInOut 0.4.7
=======================

Add functions mspct2matrix() and matrix2mspct for conversion of collections 
of spectra into R matrices and vice-versa.

photobiologyInOut 0.4.6
=======================


R/read.xlsx.r
##' Read data from .xlsx files.
##' 
##' Reads and parses an xlsx file returning a data table.
##' 
##' @usage read.xlsx(file, header = FALSE, sheets2keep = NULL)
##' @param file the name of the file which data are to be read from.
##' @param header a logical value indicating whether the file contains names of variables.
##' @param sheets2keep a vector of character strings with the names of the worksheets to be read.
##' 
##' @return Returns a data table.
##' @export
##' @author Pedro J. Aphalo, Schaun Jacob Wheeler
##' 
##' @references \url{http://www.r4photobiology.info}
##' @keywords misc
##' @note The MIT License (MIT)
##'
##' Copyright (c) 2012 Schaun Jacob Wheeler
##'
##' Permission is hereby granted, free of charge, to any person obtaining a copy
##' of this software and associated documentation files (the "Software"), to deal
##' in the Software without restriction, including without limitation the rights
##' to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
##' copies of the Software, and to permit persons to whom the Software is
##' furnished to do so, subject to the following conditions:
##'
##' The above copyright notice and this permission notice shall be included in all
##' copies or substantial portions of the Software.
##'
##' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
##' IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
##' FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
##' AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
##' LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
##' OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
##' SOFTWARE.

read.xlsx <- function(file, header = FALSE, skip = 0, sheets2keep = NULL) {

  temp_dir <- file.path(tempdir(), "read.xlsx.temp")
  suppressWarnings(dir.create(temp_dir))

  file.copy(file, temp_dir)
  new_file <- list.files(temp_dir, full.names = TRUE, pattern = basename(file))
  unzip(new_file, exdir = temp_dir)

  # Get OS
  # These lines are included because R documentation states that Excel handles
  # date origins differently on Mac than on Windows. However, manual inspection
  # of Excel files created on Windows and Mac indicated that in fact the origin
  # is handled the same across both platforms. I've kept the original code here
  # commented out in case it can be of use in the future.
  # mac <- xmlToList(xmlParse(list.files(
  #   paste0(temp_dir, "/docProps"), full.names = TRUE, pattern = "app.xml")))
  # mac <- grepl("Macintosh", mac$Application)
  # if(mac) {
  #   os_origin <- "1899-12-30" # documentation says should be "1904-01-01"
  # } else {
  #   os_origin <- "1899-12-30"
  # }

  os_origin <- "1899-12-30"
  
  # Get names of sheets
  sheet_names <- xmlToList(xmlParse(list.files(
    paste0(temp_dir, "/xl"), full.names = TRUE, pattern = "workbook.xml")))
  sheet_names <- rbind.fill(lapply(sheet_names$sheets, function(x) {
    as.data.frame(as.list(x), stringsAsFactors = FALSE)
  }))
  rownames(sheet_names) <- NULL
  sheet_names <- as.data.frame(sheet_names,stringsAsFactors = FALSE)
  sheet_names$id <- gsub("\\D", "", sheet_names$id)

  # Get column classes
  styles <- xmlParse(list.files(
    paste0(temp_dir, "/xl"), full.names = TRUE, pattern = "styles.xml"))
  styles <- xpathApply(styles, "//x:xf[@applyNumberFormat and @numFmtId]",
    namespaces = "x", xmlAttrs)
  styles <- lapply(styles, function(x) {
    x[grepl("applyNumberFormat|numFmtId", names(x))]})
  styles <- do.call("rbind", (lapply(styles,
    function(x) as.data.frame(as.list(x[c("applyNumberFormat", "numFmtId")]),
      stringsAsFactors = FALSE))))

  if(!is.null(sheets2keep)) {
    sheet_names <- sheet_names[sheet_names$name %in% sheets2keep,]

  }

  worksheet_paths <- list.files(
    paste0(temp_dir, "/xl/worksheets"),
    full.names = TRUE,
    pattern = paste0(
      "sheet(",
      paste(sheet_names$id, collapse = "|"),
      ")\\.xml$"))

  worksheets <- lapply(worksheet_paths, function(x) xmlRoot(xmlParse(x))[["sheetData"]])

  worksheets <- pblapply(seq_along(worksheets), function(i) {

    x <- xpathApply(worksheets[[i]], "//x:c", namespaces = "x", function(node) {
      c("v" = xmlValue(node[["v"]]), xmlAttrs(node))
    })

    if(length(x) > 0) {

      x_rows <- unlist(lapply(seq_along(x), function(i) rep(i, length(x[[i]]))))
      x <- unlist(x)

      x <- reshape(
        data.frame(
          "row" = x_rows,
          "ind" = names(x),
          "value" = x,
          stringsAsFactors = FALSE),
        idvar = "row", timevar = "ind", direction = "wide")

      x$sheet <- sheet_names[sheet_names$id == i, "name"]
      colnames(x) <- gsub("^value\\.", "", colnames(x))
    }
    x
  })
  worksheets <- do.call("rbind.fill",
    worksheets[sapply(worksheets, class) == "data.frame"])

  entries <- xmlParse(list.files(paste0(temp_dir, "/xl"), full.names = TRUE,
    pattern = "sharedStrings.xml$"))
  entries <- xpathSApply(entries, "//x:si", namespaces = "x", xmlValue)
  names(entries) <- seq_along(entries) - 1

  entries_match <- entries[
    match(worksheets$v[worksheets$t == "s" & !is.na(worksheets$t)],
      names(entries))]
  worksheets$v[worksheets$t == "s" & !is.na(worksheets$t)] <- entries_match
  worksheets$cols <- match(gsub("\\d", "", worksheets$r), LETTERS)
  worksheets$rows <- as.numeric(gsub("\\D", "", worksheets$r))

  if(!any(grepl("^s$", colnames(worksheets)))) {
    worksheets$s <- NA
  }

  workbook <- lapply(unique(worksheets$sheet), function(x) {
    y <- worksheets[worksheets$sheet == x,]
    y_style <- as.data.frame(tapply(y$s, list(y$rows, y$cols), identity),
      stringsAsFactors = FALSE)
    y <- as.data.frame(tapply(y$v, list(y$rows, y$cols), identity),
      stringsAsFactors = FALSE)

    if(header) {
      colnames(y) <- y[1,]
      y <- y[-1,]
      y_style <- y_style[-1,]
    }

    y_style <- sapply(y_style, function(x) {
      out <- names(which.max(table(x)))
      out[is.null(out)] <- NA
      out
    })

    if(length(styles) > 0) {
      y_style <- styles$numFmtId[match(y_style, styles$applyNumberFormat)]
    }

    y_style[y_style %in% 14:17] <- "date"
    y_style[y_style %in% c(18:21, 45:47)] <- "time"
    y_style[y_style %in% 22] <- "datetime"
    y_style[is.na(y_style) & !sapply(y, function(x)any(grepl("\\D", x)))] <- "numeric"
    y_style[is.na(y_style)] <- "character"
    y_style[!(y_style %in% c("date", "time", "datetime", "numeric"))] <- "character"

    y[] <- lapply(seq_along(y), function(i) {
      switch(y_style[i],
        character = y[,i],
        numeric = as.numeric(y[,i]),
        date = as.Date(as.numeric(y[,i]), origin = os_origin),
        time = strftime(as.POSIXct(as.numeric(y[,i]), origin = os_origin), format = "%H:%M:%S"),
        datetime = as.POSIXct(as.numeric(y[,i]), origin = os_origin))
    })
    y
  })

  if(length(workbook) == 1) {
    workbook <- workbook[[1]]
  } else {
    names(workbook) <- sheet_names$name
  }

  workbook
}
